### PR TITLE
feat: Add virtual machine creation wizard

### LINKED ISSUE #19: Virtual machine builder
One of the most complicated parts of the CLI is going to be the UI for building a new virtual machine. This needs to be really nice to use. 

Building a virtual machine is bloody complicated and has lots of options that the user can configure. Whether we want to support all of these in the UI will come down to striking a balance between complexity and flexibility. 

To begin, we should focus on an interactive mode for creating VMs. The mode will need to ask a series of questions (some of which having selections, some text input etc...), the build an appropriate [build spec](https://gist.github.com/adamcooke/a4e25364db62ea9bcd7a9997c63d82f5) and ship that off to the API. 

(It would actually be *really* nice to have a mode in the CLI that runs through the wizard but just spits out the build spec at the end rather than actually sending it to the API. Some sort of "preview".)

The question that need to be asked are likely to be very similar to the flow that exists in the current UI and of the new UI.

Screenshot

To simplify this for an initial version, let's do the following:

1. Choose a data center (basically a list to choose from which will need to be looked up in the API)

2. Choose a package (they need to choose which package to use, there are lots of packages so this may need some clever fuzzy searching to allow searching for a specific size). We also need to support flexible resources. I'd suggest we have a flag that can be passed to the command (`--flexible`) which will instruct it to ask for a quantity of memory and CPU cores rather than even use the package selector.

  Which ever option is chosen, we also need to ask for a size for the system disk.

3. Choose a template (as above, there could be quite a few so needs a good fuzzy search). There should also be an option to use no template.

4. Ask for hostname, name, group, description, tags etc...

There are other things we can do which will need to be integrated eventually somehow. We ultimately want to support using any part of the build spec (linked above).

* Configuring network interfaces
* Choosing which SSH keys to allow
* Selecting ISOs instead of templates

### PR DESCRIPTION
**Pull request by JakeMakesStuff**
fixes #19

### CODE REVIEW

File: cmd/katapult/console/selector.go [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+func (gotermTerminal) MakeRaw() (*term.State, error) {
+	return term.MakeRaw(0)
+}
+
+// items is either []string or [][]string (if columns isn't nil).
+//nolint:funlen
+func selectorComponent(
+	question string, columns []string, items interface{},
```
Comment (strideynet) [inline cmd/katapult/console/selector.go:337]: I think I prefer one per line, this feels hard to read/scan.

**Review by strideynet — CHANGES_REQUESTED** [review]
Add some tests for that creation wizard !!

File: cmd/katapult/console/selector.go [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+	}
+	return x
+}
+
+func exactCompare(a, b interface{}) bool {
+	switch x := a.(type) {
+	case []string:
+		// The logic behind this: A string slice is just a fancy pointer with some extra data.
+		// If we compare the pointers, we can tell if they are equal.
+		return &x[0] == &b.([]string)[0]
```
Comment (strideynet) [inline cmd/katapult/console/selector.go:48]: This feels very panicky. Are we sure zero length arrays won't be passed in, and that B will always be the same type?

File: cmd/katapult/console/selector.go [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+			matched = [][]string{}
+		} else {
+			matched = []string{}
+		}
+	}
+	return matched, query, matchedLen
+}
+
+// Formats the user prompt. Returns the rough line count.
+func formatUserPrompt(length, highlightIndex int, hasColumns bool, matched interface{},
```
Comment (strideynet) [inline cmd/katapult/console/selector.go:102]: This wrapping is pretty weird, can we have one parameter per line maybe ?

File: cmd/katapult/console/selector.go [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+		if len(column) >= lengthPerColumn {
+			// Truncate
+			content += column[:lengthPerColumn]
+		} else {
+			// Pad
+			content += column + strings.Repeat(" ", lengthPerColumn-len(column))
+		}
+	}
+	switch highlight {
+	case 1:
```
Comment (strideynet) [inline cmd/katapult/console/selector.go:159]: Could we use some const values for these?
```go
const (
  HighlightNone = 0
  HighlightContent = 1
  HighlightTitle = 2
)
```

File: cmd/katapult/console/selector.go (hunk 2 above)
Comment (JakeMakesStuff) [reply cmd/katapult/console/selector.go:48]: Yes. It is only called when items are selected, which would mean items exist in the first place. 👍

**Review by strideynet — APPROVED** [review]

**Pull request merged**